src/main/java/com/google/devtools/build/lib/buildtool/ExecutionTool.java

```diff
@@ -766,16 +766,35 @@ private ImmutableList<ConvenienceSymlink> createConvenienceSymlinks(
     Reporter reporter = env.getReporter();
 
     // Gather configurations to consider.
-    Set<BuildConfigurationValue> targetConfigurations =
-        buildRequestOptions.useTopLevelTargetsForSymlinks() && !targetsToBuild.isEmpty()
-            ? targetsToBuild.stream()
-                .map(ConfiguredTarget::getActual)
-                .map(ConfiguredTarget::getConfigurationKey)
-                .filter(Objects::nonNull)
-                .distinct()
-                .map((key) -> executor.getConfiguration(reporter, key))
-                .collect(toImmutableSet())
-            : ImmutableSet.of(configuration);
+    ImmutableSet<BuildConfigurationValue> targetConfigs;
+    if (buildRequestOptions.useTopLevelTargetsForSymlinks() && !targetsToBuild.isEmpty()) {
+      // Collect the configuration of each top-level requested target. These may be different than
+      // the build's top-level configuration because of self-transitions.
+      ImmutableSet<BuildConfigurationValue> requestedTargetConfigs =
+          targetsToBuild.stream()
+              .map(ConfiguredTarget::getActual)
+              .map(ConfiguredTarget::getConfigurationKey)
+              .filter(Objects::nonNull)
+              .distinct()
+              .map((key) -> executor.getConfiguration(reporter, key))
+              .collect(toImmutableSet());
+      if (requestedTargetConfigs.size() == 1) {
+        // All top-level targets have the same configuration, so use that one.
+        targetConfigs = requestedTargetConfigs;
+      } else if (requestedTargetConfigs.contains(configuration)) {
+        // Mixed configs but at least one of them includes the top-level config. Set symlinks to the
+        // top-level config so at least non-transitioned targets resolve. See
+        // https://github.com/bazelbuild/bazel/issues/17081.
+        targetConfigs = ImmutableSet.of(configuration);
+      } else {
+        // Mixed configs, none of which include the top-level config. Delete the symlinks because
```

Comment:
Should we include this case? 

Pro: the symlink has nothing useful
Con: this will trigger when `$ bazel build //...` doesn't include any tests, since `--trim_test_configuration` will remove test options from all targets. Users may be confused why there are no symlinks.

Reply:
Could you elaborate on the Con? Does `--trim_test_configuration` result in duplicate configs? If so, could we somehow recognize when a config differs from the top-level config only in test options and ignore those configs here to fix the Con?

Reply:
It's just a con in that it means `bazel-bin` will still be deleted more than users expect. If `$ bazel build //:all` expands to 10 binaries, where only 1 has a self-transition, `bazel-bin` will still be deleted. If `$ bazel build //:all` also includes a test, `bazel-bin` will show up. 

Worse, the non-transitioning binaries would actually appear in `bazel-bin` if it wasn't deleted. Even though they lack `TestOptions` they still have the same `bazel-out/foo-bar/bin` path because that instance of config trimming doesn't affect paths.

I'd rather not get into explicitly examining the configuration diffs. Another approach is to not compare the configs but their output paths. 

Latest commit implements that, and manual testing shows intuitive results (I tried every combo of test, non-transitioning non-test, transitioning non-test).

Reply:
I like that change, comparing paths instead of configs makes sense as those are all we care about in the end.